var/spack/repos/builtin/packages/py-mysqlclient/package.py

```diff
@@ -17,10 +17,17 @@ class PyMysqlclient(PythonPackage):
 
     license("GPL-2.0-or-later")
 
+    version("2.2.4", sha256="33bc9fb3464e7d7c10b1eaf7336c5ff8f2a3d3b88bab432116ad2490beb3bf41")
     version("1.4.6", sha256="f3fdaa9a38752a3b214a6fe79d7cae3653731a53e577821f9187e67cbecb2e16")
     version("1.4.5", sha256="e80109b0ae8d952b900b31b623181532e5e89376d707dcbeb63f99e69cefe559")
     version("1.4.4", sha256="9c737cc55a5dc8dd3583a942d5a9b21be58d16f00f5fefca4e575e7d9682e98c")
     version("1.3.13", sha256="ff8ee1be84215e6c30a746b728c41eb0701a46ca76e343af445b35ce6250644f")
 
     depends_on("py-setuptools", type="build")
     depends_on("mysql")
+
+    # Fix "library not found for -lzlib" as described in
+    # https://github.com/PyMySQL/mysqlclient/issues/584 by backporting the
+    # patch from @2.2.0rc1 that uses pkg-config instead of mysql_config
+    # https://github.com/PyMySQL/mysqlclient/pull/586/files
+    patch("use-pkg-config.patch", when="@:2.1")
```

Comment:
We prefer to use upstream urls to patches when available (for reasons of copyright, attribution, provenance), e.g.
```suggestion
    patch(
        "https://github.com/PyMySQL/mysqlclient/pull/586.patch?full_index=1",
        sha256="8cedca44c93606efc64ed6bf87ffed24e0da4841d2c28aacf760e9bd833d1967"
        when="@:2.1",
    )
```
This will also pull in the fix for `setup_windows.py`.

Can you confirm that this patch applies to all versions? Or should it be `when="@2:2.1"`?

Reply:
Thanks for the suggestions of using the PR patch. The PR patch doesn't work for any of the prior versions. My patch works for the 1.4 versions onwards (but not 1.3.13) so the patch could be revised to `@1:4:2.1`. I have no way of testing whether the windows patch works and am not keen to volunteering my time to support a proprietary operating system. Below are the rejections from using the upstream patch:

```console
$ grep -hB6 -F 'Apply anyway?' 02-error/slurm-23673032_*
==> Installing py-mysqlclient-1.4.6-nsj5ugli2y73bylfji56as2stw5ljwgd [38/38]
==> No binary for py-mysqlclient-1.4.6-nsj5ugli2y73bylfji56as2stw5ljwgd found: installing from source
1 out of 1 hunk FAILED -- saving rejects to file metadata.cfg.rej
1 out of 1 hunk FAILED -- saving rejects to file setup_common.py.rej
2 out of 2 hunks FAILED -- saving rejects to file setup_posix.py.rej
Reversed (or previously applied) patch detected!  Assume -R? [n]
Apply anyway? [n]
--
==> Installing py-mysqlclient-1.4.5-y66a4gfmxhqwb762hisdyjlu2s7b3k6u [38/38]
==> No binary for py-mysqlclient-1.4.5-y66a4gfmxhqwb762hisdyjlu2s7b3k6u found: installing from source
1 out of 1 hunk FAILED -- saving rejects to file metadata.cfg.rej
1 out of 1 hunk FAILED -- saving rejects to file setup_common.py.rej
2 out of 2 hunks FAILED -- saving rejects to file setup_posix.py.rej
Reversed (or previously applied) patch detected!  Assume -R? [n]
Apply anyway? [n]
--
==> Installing py-mysqlclient-1.4.4-3vwqtngyvkarhlvwgdbrc2vai2j2b7qm [38/38]
==> No binary for py-mysqlclient-1.4.4-3vwqtngyvkarhlvwgdbrc2vai2j2b7qm found: installing from source
1 out of 1 hunk FAILED -- saving rejects to file metadata.cfg.rej
1 out of 1 hunk FAILED -- saving rejects to file setup_common.py.rej
2 out of 2 hunks FAILED -- saving rejects to file setup_posix.py.rej
Reversed (or previously applied) patch detected!  Assume -R? [n]
Apply anyway? [n]
--
==> Installing py-mysqlclient-1.3.13-tzvifi7put66hh3fngwl4pbg6qpjlnor [38/38]
==> No binary for py-mysqlclient-1.3.13-tzvifi7put66hh3fngwl4pbg6qpjlnor found: installing from source
1 out of 1 hunk FAILED -- saving rejects to file metadata.cfg.rej
1 out of 1 hunk FAILED -- saving rejects to file setup_common.py.rej
2 out of 2 hunks FAILED -- saving rejects to file setup_posix.py.rej
Reversed (or previously applied) patch detected!  Assume -R? [n]
Apply anyway? [n]
2 out of 2 hunks ignored -- saving rejects to file setup_windows.py.rej
1 out of 1 hunk FAILED -- saving rejects to file README.md.rej
Reversed (or previously applied) patch detected!  Assume -R? [n]
Apply anyway? [n]
```

Reply:
Thanks. I see the problem with the patch by pull request URL: the pull request also updates the version number in the `metdata.cfg` file so it will only apply cleanly against a single version :-\

In that case, let's go with what you suggest and simply change the patch when range to `@1:4:2.1`. Windows users will (presumably) be able to use the 2.2.4 version you added.

Reply:
I do appreciate your testing, of course. And I understand the hesitation to spending too much effort on an operating system that you cannot test on (I'm in the same situation and have to rely on the spack CI builds on windows for that).

Reply:
New suggestion, for your convenience.
```suggestion
    patch("use-pkg-config.patch", when="@1.4:2.1")
```